taskvine/src/manager/vine_manager.c

```diff
@@ -4048,6 +4005,24 @@ int vine_set_password_file(struct vine_manager *q, const char *file)
 	return copy_file_to_buffer(file, &q->password, NULL) > 0;
 }
 
+static void delete_task_at_exit(struct vine_task *t)
+{
+	if (!t) {
+		return;
+	}
+
+	if (t->type == VINE_TASK_TYPE_LIBRARY) {
```

Comment:
Hmm, this doesn't seem right to me.
I think our general principle should be: whoever added a reference is responsible for deleting that reference.
e.g. If a user created a task once, then they are responsible for deleting it once.  If the manager cloned it once, then the manager is responsible for deleting it once.  Just deleting all available references will cause trouble if there is some other entity that had a reason to clone the object.

Reply:
In this case the manager is the user. The manager `vine_task_copy` these tasks. This case is handling library tasks that are never DONE. If they were DONE, they would be deleted by the manager.

Reply:
One thing that could work is to re-add the slot check to that function, and just try to send the library at the last step. If the library does not fit, then it shouldn't be sent again.

Reply:
That worked, tested on coffea-casa.

Reply:
One alternative here is to not `vine_task_clone` library tasks when adding them to the task table, and then not `vine_task_delete` them when they are DONE.

Reply:
Ah, I think this is the source of the confusion:

The `q->library` table contains *only* the prototype library tasks that are submitted by the user via `install_library`. So the user has one reference, and then the manager adds one when adding it to the `library` table. So, the manager should simple remove one reference when taking it out of the table.  (Maybe that table should have a more precise time.)

Now, when it is time to deploy a library to a worker, the manager **copies** the prototype task, and then puts it into the `tasks` table with an additional reference.  (Just like any other task.). So, when the library task is done, it should lose a reference for being taking out of the task table (just like any other task).  And then you are right, the manager (as creator) of that task should delete its final reference.

I think we are getting mixed up between the prototype library tasks and their copies.  Perhaps we need a better naming scheme to make it clear.

Reply:
Since the original is copied, I think that the copies also get the `LIBRARY` type?
The originals were also mem-leaked, but that was fixed in another pr.

Since this is not fixing the mem leak anyway, I can remove it from the pr and merge the rest that we know works.